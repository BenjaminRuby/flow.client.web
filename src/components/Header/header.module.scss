@import "~@boomerang/boomerang-styles/lib/colors";

.container {
  align-items: flex-start;
  background: $white;
  display: flex;
  flex-direction: column;
  padding: 1rem 2rem 2rem;
  position: sticky;
  top: 0;
  width: 100%;
  z-index: 3; //min needed, 99 puts is above the sidenav
}
.content {
  align-items: center;
  display: flex;
  justify-content: space-between;
  width: 100%;
}

.info {
  overflow: hidden;
}

.title {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  line-height: 2.5rem;
}

.description {
  font-size: 1rem;
  line-height: 1.5;
  overflow: hidden;
  text-overflow: wrap;
}

.buttons {
  display: flex;
  justify-content: space-between;
  flex-flow: row wrap;
  button {
    margin-top: 1rem;
  }
}
