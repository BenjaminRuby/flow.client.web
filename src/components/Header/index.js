import React from "react";
import PropTypes from "prop-types";
import styles from "./header.module.scss";

Header.propTypes = {
  title: PropTypes.string.isRequired,
  description: PropTypes.string,
  Button: PropTypes.func.isRequired
};

function Header({ title, description, Button }) {
  return (
    <div className={styles.container}>
      <div className={styles.content}>
        <div className={styles.info}>
          <h1 className={styles.title}>{title}</h1>
          {description && <p className={styles.description}>{description}</p>}
        </div>
        <section className={styles.buttons}>
          <Button />
        </section>
      </div>
    </div>
  );
}

export default Header;
