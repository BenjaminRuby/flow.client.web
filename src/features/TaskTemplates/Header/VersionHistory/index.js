import React from "react";
import PropTypes from "prop-types";
import {
  Button,
  ComposedModal,
  ModalBody,
  StructuredListWrapper,
  StructuredListHead,
  StructuredListBody,
  StructuredListRow,
  StructuredListCell,
} from "@boomerang/carbon-addons-boomerang-react";
import { Catalog16 } from "@carbon/icons-react";
import styles from "./versionHistory.module.scss";

VersionHistory.propTypes = {
  revisions: PropTypes.array,
};

function VersionHistory({ changelogs }) {
  return (
    <ComposedModal
      composedModalProps={{ shouldCloseOnOverlayClick: true }}
      modalHeaderProps={{
        title: "Version History",
        subtitle: "Check this task template changelog.",
      }}
      modalTrigger={({ openModal }) => (
        <Button
          onClick={openModal}
          renderIcon={Catalog16}
          className={styles.history}
          kind="ghost"
          tooltipPosition="right"
          tooltipAlignment="center"
          iconDescription="History"
          size="small"
          hasIconOnly
        />
      )}
    >
      {() => <VersionHistoryModal changelogs={changelogs} />}
    </ComposedModal>
  );
}

function VersionHistoryModal({ changelogs }) {
  return (
    <ModalBody>
      <StructuredListWrapper>
        <StructuredListHead>
          <StructuredListRow head>
            <StructuredListCell head>Version</StructuredListCell>
            <StructuredListCell head>Creator</StructuredListCell>
            <StructuredListCell head>Comments</StructuredListCell>
            <StructuredListCell head>Date</StructuredListCell>
          </StructuredListRow>
        </StructuredListHead>
        <StructuredListBody>
          {changelogs.map((log) => (
            <StructuredListRow>
              <StructuredListCell>{log?.version ?? "--"}</StructuredListCell>
              <StructuredListCell noWrap>{log?.userName ?? "--"}</StructuredListCell>
              <StructuredListCell>{log?.reason ?? "--"}</StructuredListCell>
              <StructuredListCell noWrap>{log?.date ?? "--"}</StructuredListCell>
            </StructuredListRow>
          ))}
        </StructuredListBody>
      </StructuredListWrapper>
    </ModalBody>
  );
}

export default VersionHistory;
