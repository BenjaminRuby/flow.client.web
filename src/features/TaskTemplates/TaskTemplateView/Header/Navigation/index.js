import React from "react";
import PropTypes from "prop-types";
import { NavLink, withRouter } from "react-router-dom";
import styles from "./Navigation.module.scss";

Navigation.propTypes = {
  location: PropTypes.object.isRequired,
  match: PropTypes.object.isRequired
};

function Navigation({ location, match }) {
  return (
    <nav className={styles.links}>
      <NavLink exact className={styles.link} activeClassName={styles["--active"]} to={`${match.url}`}>
        Overview
      </NavLink>
      <NavLink className={styles.link} activeClassName={styles["--active"]} to={`${match.url}/settings`}>
        Settings
      </NavLink>
    </nav>
  );
}

export default withRouter(Navigation);
