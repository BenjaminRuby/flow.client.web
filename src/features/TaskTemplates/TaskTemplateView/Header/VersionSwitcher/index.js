import React, { Component } from "react";
import PropTypes from "prop-types";
import cx from "classnames";
import { Prompt } from "react-router-dom";
import { ChevronLeft16, ChevronRight16, PageFirst16, PageLast16 } from "@carbon/icons-react";
import styles from "./VersionSwitcher.module.scss";

class VersionSwitcher extends Component {
  static propTypes = {
    currentRevision: PropTypes.number,
    setCurrentRevision: PropTypes.func,
    revisionCount: PropTypes.number.isRequired,
    onChangeVersion: PropTypes.func.isRequired,
    revisions: PropTypes.array
  };

  backVersion = () => {
    this.props.setCurrentRevision(this.props.revisions[this.props.currentRevision.version - 2]);
  };

  fastBackVersion = () => {
    this.props.setCurrentRevision(this.props.revisions[0]);
  };

  forwardVersion = () => {
    this.props.setCurrentRevision(this.props.revisions[this.props.currentRevision.version]);
  };

  fastForwardVersion = () => {
    this.props.setCurrentRevision(this.props.revisions[this.props.revisions.length - 1]);
  };

  renderBackButtons = enabled => {
    return (
      <div className={styles.buttonList}>
        <button className={styles.button} disabled={!enabled} onClick={this.fastBackVersion}>
          <PageFirst16 className={cx(styles.icon, { [styles.disabled]: !enabled })} alt="first version" />
        </button>
        <button className={styles.button} disabled={!enabled} onClick={this.backVersion}>
          <ChevronLeft16 className={cx(styles.icon, { [styles.disabled]: !enabled })} alt="back one version" />
        </button>
      </div>
    );
  };

  renderForwardButtons = enabled => {
    return (
      <div className={styles.buttonList}>
        <button className={styles.button} disabled={!enabled} onClick={this.forwardVersion}>
          <ChevronRight16 className={cx(styles.icon, { [styles.disabled]: !enabled })} alt="forward one version" />
        </button>
        <button className={styles.button} disabled={!enabled} onClick={this.fastForwardVersion}>
          <PageLast16 className={cx(styles.icon, { [styles.disabled]: !enabled })} alt="last version" />
        </button>
      </div>
    );
  };

  render() {
    const { currentRevision, disabled, revisionCount, isDirty } = this.props;
    return (
      <div className={cx(styles.container, { [styles.disabled]: disabled })}>
        <div className={styles.buttonListContainer}>
          {this.renderBackButtons(currentRevision.version > 1)}
          <p className={styles.versionText}>{`Version ${currentRevision?.version || 1}`}</p>
          {this.renderForwardButtons(currentRevision.version < revisionCount)}
        </div>
      </div>
    );
  }
}

export default VersionSwitcher;
