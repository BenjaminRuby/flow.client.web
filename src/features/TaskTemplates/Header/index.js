import React from "react";
import { useHistory } from "react-router-dom";
import { Button } from "@boomerang/carbon-addons-boomerang-react";
import capitalize from "lodash/capitalize";
import FeatureHeader from "Components/FeatureHeader";
import VersionSwitcher from "./VersionSwitcher";
import { serviceUrl } from "Config/servicesConfig";
import { Bee20, Save16, Undo16, WatsonHealthSaveSeries16 } from "@carbon/icons-react";
import styles from "./header.module.scss";

function Header({ selectedTaskTemplate, currentRevision }) {
  let history = useHistory();
  return (
    <FeatureHeader includeBorder className={styles.featureHeader}>
      <div className={styles.container}>
        <div>
          <h1 className={styles.category}>{capitalize(selectedTaskTemplate.category)}</h1>
          <div className={styles.infoContainer}>
            {selectedTaskTemplate.imageId ? (
              <img
                className={styles.icon}
                alt={`${selectedTaskTemplate.name} icon`}
                src={serviceUrl.getImage({imgId:selectedTaskTemplate.imageId})}
              />
            ) : (
              <Bee20 alt={`${selectedTaskTemplate.name} icon`} className={styles.icon} />
            )}
            <p className={styles.taskName}>{selectedTaskTemplate.name || "TASK"}</p>
            {/* <AddService task={task} /> */}
          </div>
        </div>
        <div className={styles.buttons}>
          <VersionSwitcher 
            revisions={selectedTaskTemplate.revisions} 
            currentRevision={currentRevision} 
            revisionCount={selectedTaskTemplate.revisions.length}
          />
          <Button className={styles.button} size="field" kind="ghost" renderIcon={Undo16}>Reset changes</Button>
          {/* <Button className={styles.button} size="field" kind="secondary" renderIcon={Save16}>Update this version</Button> */}
          <Button className={styles.button} size="field" renderIcon={WatsonHealthSaveSeries16}>Save new version</Button>
        </div>
      </div>
    </FeatureHeader>
  );
}

export default Header;
