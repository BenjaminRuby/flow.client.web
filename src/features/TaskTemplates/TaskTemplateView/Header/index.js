import React from "react";
import PropTypes from "prop-types";
import { Link, useHistory } from "react-router-dom";
import { useMutation } from "react-query";
import { ConfirmModal, notify, ToastNotification, Loading, Button } from "@boomerang/carbon-addons-boomerang-react";
import FeatureHeader from "Components/FeatureHeader";
import Navigation from "./Navigation";
import VersionSwitcher from "./VersionSwitcher";
import { DocumentExport16 } from "@carbon/icons-react";
import { resolver } from "Config/servicesConfig";
import { QueryStatus } from "Constants/reactQueryStatuses";
import styles from "./Header.module.scss";

Header.propTypes = {
  currentRevision: PropTypes.number,
  fetchTaskTemplateVersion: PropTypes.func,
  handleChangeLogReasonChange: PropTypes.func,
  onDesigner: PropTypes.bool.isRequired,
  performAction: PropTypes.func,
  performActionButtonText: PropTypes.string,
  revisionCount: PropTypes.number,
  taskTemplateName: PropTypes.string.isRequired
};
Header.defaultProps = {
  includeResetVersionAlert: false
};

function Header ({
  loading,
  isValid,
  currentRevision,
  revisionCount,
  taskTemplateToEdit,
  revisions,
  values,
  isDirty,
  isEdit,
  setSubmitting
}) {
  const [CreateTaskTemplateMutation, {status: createTaskTemplateStatus}] = useMutation(resolver.postCreateTaskTemplate);
  const [UploadTaskTemplateMutation, {status: updateTaskTemplateStatus}] = useMutation(resolver.putCreateTaskTemplate);
  const isLoadingCreate = createTaskTemplateStatus === QueryStatus.Loading;
  const isLoadingUpdate = updateTaskTemplateStatus === QueryStatus.Loading;
  const history = useHistory();

  const handleSubmitTaskTemplate = async () => {
    const newRevisions = [].concat(revisions??[]);
    // const currentRevisionIndex = !isEdit? 0 : revisions.findIndex(revision => revision.version === currentRevision.version);
    let newRevisionConfig = {
      version: Boolean(currentRevision?.version) ? currentRevision?.version + 1 : 1,
      image: values.image, 
      // image: "container:version", 
      command: values.command,
      // command: "bmrgctl",
      arguments : values.arguments.trim().split(/\s{1,}/),
      config: values.settings
    };
    newRevisions.push(newRevisionConfig);
    const body = {
      ...taskTemplateToEdit,
      name: values.name,
      description: values.description,
      category: values.category,
      key: values.key,
      revisions:newRevisions,
      nodetype: "templateTask"
    };
    if(!isEdit){
      try {
        await CreateTaskTemplateMutation({ body: body });
        notify(
          <ToastNotification
            kind="success"
            title={"Task Template Created"}
            subtitle={`Request to create ${body.name} succeeded`}
            data-testid="create-update-task-template-notification"
          />
        );
        setSubmitting(true);
        history.push("/task-templates");
      } catch (err) {
        notify(
          <ToastNotification
            kind="error"
            title={"Create Task Template Failed"}
            subtitle={"Something's Wrong"}
            data-testid="create-update-task-template-notification"
          />
        );
    }
  }
    else{
      try {
        await UploadTaskTemplateMutation({ body });
        notify(
          <ToastNotification
            kind="success"
            title={"Task Template Updated"}
            subtitle={`Request to update ${body.name} succeeded`}
            data-testid="create-update-task-template-notification"
          />
        );
      } catch (err) {
        notify(
          <ToastNotification
            kind="error"
            title={"Update Task Template Failed"}
            subtitle={"Something's Wrong"}
            data-testid="create-update-task-template-notification"
          />
        );
      }
    }
  };
  const determinePerformActionRender = () => {
    const performActionButtonText=isEdit? "Create New Version" : "Create";
    const message = isEdit? `Version ${currentRevision.version + 1} will be created and users will be prompted to update their workflows.`: `The first version of ${values.name} will be created and available for use in workflows.`;
    return (
        <ConfirmModal
          affirmativeAction={handleSubmitTaskTemplate}
          children={message}
          title={`${isEdit?"Create":"Create new version"} - ${values.name}`}
          modalTrigger={({ openModal }) => (
            <Button
              disabled={loading || !isValid || !isDirty}
              iconDescription="Set version to latest"
              kind="ghost"
              onClick={openModal}
              renderIcon={DocumentExport16}
              size="field"
            >
              {performActionButtonText}
            </Button>
          )}
        />
    );
  }
    return (
      <FeatureHeader includeBorder className={styles.container}>
        {(isLoadingCreate || isLoadingUpdate) && <Loading />}
        <section className={styles.header}>
          <div className={styles.breadcrumbContainer}>
            <Link className={styles.workflowsLink} to="/task-templates">
              Task Templates
            </Link>
            <span className={styles.breadcrumbDivider}>/</span>
            <p className={styles.taskTemplateName}> {taskTemplateToEdit?.name?? "New Task Template"}</p>
          </div>
          <h1 className={styles.title}>Task Template</h1>
        </section>
        <section className={styles.versionButtons}>
          <VersionSwitcher
            currentRevision={currentRevision}
            revisionCount={revisionCount}
            revisions={revisions}
            isDirty={isDirty}
            settings={values.settings}
          />
          {determinePerformActionRender()}
        </section>
        <Navigation />
      </FeatureHeader>
    );
}

export default Header;
