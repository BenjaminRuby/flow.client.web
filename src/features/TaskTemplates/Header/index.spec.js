import React from "react";
import { wait } from "@testing-library/react";
import AppContext from "state/context/appContext";
import Header from "./index";

const serviceDetails = {
  id: "5b5b451536c18900018a33d1",
  icon: null,
  name: "Jenkins",
  category: "Mobile",
  version: "1.0.9",
  imageId: "test",
  documentationUrl: "http://doc.url.com",
  createdDate: null,
  dateLastUpdated: "2019-11-11T02:47:59.719+0000",
  attributes: ["Associated cost", "Third party", "Admin-only"],
  heroImage: {
    id: "1",
    filename: "name",
    description: "about me",
    alt: "shiba",
    label: "Test Img Shiba",
    src: "https://encrypted-tbn0.gstatic.com/images?q=tbn%3AANd9GcSiYlHMBwjiaaGOYD_bNdUoHU8tB58jqO9ycWS2UyIXoQ0avTdF"
  },
  productImages: [
    {
      imageId: "1",
      name: "Test Img Panda",
      src: "https://cdnbr2.img.sputniknews.com/images/1395/55/13955519.jpg",
      description: "about me"
    },
    {
      imageId: "2",
      name: "Test Img Red Panda",
      src: "https://static.hitek.fr/img/actualite/26rrbijj.jpg",
      description: "about me",
      url: "https://media.gettyimages.com/photos/giant-panda-picture-id483807166?s=2048x2048"
    },
    {
      imageId: "3",
      name: "Test Img Ferret",
      src:
        "https://media.gettyimages.com/photos/giant-panda-baby-cub-in-chengdu-area-china-picture-id539113690?s=2048x2048"
    },
    {
      imageId: "4",
      name: "Test Img Kitten",
      src: "https://encrypted-tbn0.gstatic.com/images?q=tbn%3AANd9GcSJ7PlfTZfXVNCUj18TwsBb3p6Ul6D0sdjRL3SC1AJTWcXhPucr",
      description: "about me"
    }
  ],
  description:
    "### Summary\nAs a business owner or manager, the decision to opt for offshore software development can be difficult. As anyone who has experience knows, there are advantages and disadvantages to outsourcing software development.\n### Detail\nAs a business owner or manager, the decision to opt for offshore software development can be difficult. As anyone who has experience knows, there are advantages and disadvantages to outsourcing software development. How then, do you balance the risks and rewards and come out on top? Here are five tips to ensure that you receive an excellent return on investment.\nAs a business owner or manager, the decision to opt for offshore software development can be difficult. As anyone who has experience knows, there are advantages and disadvantages to outsourcing software development. How then, do you balance the risks and rewards and come out on top? Here are five tips to ensure that you receive an excellent return on investment.\n- List item\n - List item \n - List item",
  properties: [
    {
      required: true,
      placeholder: null,
      helperText: null,
      language: null,
      disabled: null,
      defaultValue: null,
      value: "test",
      values: null,
      readOnly: false,
      id: "a8debe00-3611-431d-9cef-b7bd0296f40d",
      description: "test",
      key: "test",
      label: "test",
      type: "text",
      minValueLength: null,
      maxValueLength: null,
      options: null
    },
    {
      required: true,
      placeholder: null,
      helperText: null,
      language: null,
      disabled: null,
      defaultValue: null,
      value: "a",
      values: null,
      readOnly: false,
      id: "20200f76-e676-4737-b51b-1b5dec303060",
      description: "a",
      key: "a",
      label: "a",
      type: "text",
      minValueLength: null,
      maxValueLength: null,
      options: null
    }
  ]
};
const teams = [
  {
    id: "5b5b451536c18900018a33d1",
    lowLevelGroupId: "124",
    name: "Marcus Dev Team",
    shortName: null,
    owners: [
      {
        ownerId: "59aebd0b7424530fce952fdd",
        ownerEmail: "mdroy@us.ibm.com",
        ownerName: "Marcus Roy"
      },
      {
        ownerId: "5bbcf419ee763e00011f5c03",
        ownerEmail: "gchickma@us.ibm.com",
        ownerName: "Glen Hickman"
      },
      {
        ownerId: "59aebd0c7424530fce952fde",
        ownerEmail: "trbula@us.ibm.com",
        ownerName: "Timothy Bula"
      }
    ],
    purpose: "Marcus Dev Team",
    dateCreated: 1551910631869,
    isActive: true,
    newRelicRestApiKey: null,
    newRelicQueryKey: null,
    newRelicAccountId: null,
    requestedTools: [],
    autoApproveRequests: true,
    tools: [
      {
        id: "5c8046ec329e17000178962e",
        name: "Scorecard",
        templateId: "5994b744189c33ed8433d9fe",
        url: "https://wdc3.cloud.boomerangplatform.net/dev/apps/ci",
        dateCreated: 1551910636308,
        lowerLevelGroupId: "5c8046ec329e170001789629",
        ldapGroups: ["boomerang-marcus-dev-team"],
        lowLevelGroupName: "Team Tools",
        toolTemplate: {
          id: "5994b744189c33ed8433d9fe",
          name: "Scorecard",
          imageSrc: "scorecard",
          baseUrl: "https://wdc3.cloud.boomerangplatform.net/dev/apps/scorecard",
          type: "both",
          comments: "Boomerang Scorecard",
          dateLastUpdated: 1513696633712,
          version: "2.0",
          ldapPrefix: "boomerang",
          quota: null,
          isLicensed: false,
          isActive: true,
          requiresVPN: true,
          isExternalHosted: false,
          isExternalUserAccessible: false,
          hasQuota: false,
          hasSubSelections: false,
          hasPostOnboardAction: true,
          isAdminOnly: false,
          subSelections: [],
          validRoles: ["lead", "user"],
          defaultRole: "user"
        }
      }
    ]
  }
];
describe("Header --- Snapshot", () => {
  it("Capturing Snapshot of Header", async () => {
    const { baseElement, getByText } = global.renderWithRouter(
      <AppContext.Provider value={{ user: { teams, id: "59aebd0b7424530fce952fdd" } }}>
        <Header service={serviceDetails} />
      </AppContext.Provider>
    );
    await wait(() => expect(getByText(/Add to team/i)).toBeInTheDocument());
    expect(baseElement).toMatchSnapshot();
  });
});
