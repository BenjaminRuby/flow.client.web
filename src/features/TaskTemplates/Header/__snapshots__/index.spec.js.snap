// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header --- Snapshot Capturing Snapshot of Header 1`] = `
<body
  id="app"
>
  <div>
    <header
      class="container border"
    >
      <div
        class="container"
      >
        <nav
          aria-label="Breadcrumb"
          class="breadcrumb"
        >
          <ol
            class="bx--breadcrumb bx--breadcrumb--no-trailing-slash"
          >
            <button
              class="backButton"
            >
              Catalog
            </button>
            <span
              style="margin: 0px 0.5rem;"
            >
              /
            </span>
            <li
              class="bx--breadcrumb-item bx--breadcrumb-item--current"
            >
              <a
                class="bx--link"
                href="#"
              >
                Jenkins
              </a>
            </li>
          </ol>
        </nav>
        <div
          class="infoContainer"
        >
          <div
            class="info"
          >
            <img
              alt="Jenkins icon"
              class="icon"
              src="undefined/catalog/image?id=test"
            />
            <h1
              class="title"
            >
              Jenkins
            </h1>
          </div>
          <button
            class="trigger bx--btn bx--btn--field bx--btn--ghost"
            tabindex="0"
            type="button"
          >
            Add to team
            <svg
              aria-hidden="true"
              aria-label="Add service"
              class="bx--btn__icon"
              focusable="false"
              height="20"
              preserveAspectRatio="xMidYMid meet"
              role="img"
              style="will-change: transform;"
              viewBox="0 0 32 32"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M17 15L17 8 15 8 15 15 8 15 8 17 15 17 15 24 17 24 17 17 24 17 24 15z"
              />
            </svg>
          </button>
        </div>
      </div>
    </header>
  </div>
  <div
    class="ReactModalPortal"
  />
</body>
`;
